packages/clay-data-provider/src/useResource.ts

```diff
@@ -0,0 +1,236 @@
+/**
+ * © 2019 Liferay, Inc. <https://liferay.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {
+	FetchPolicy,
+	IDataProvider,
+	NetworkStatus,
+	SYMBOL_ORIGIN,
+	TVariables,
+} from './types';
+import {timeout} from './util';
+import {useCache} from './useCache';
+import {useDebounce} from '@clayui/shared';
+import {useEffect, useRef, useState} from 'react';
+
+interface IResource extends IDataProvider {
+	onNetworkStatusChange?: (status: NetworkStatus) => void;
+}
+
+const getQueryParams = (variables: any): string | null => {
+	if (!variables) {
+		return null;
+	}
+
+	const keys = Object.keys(variables);
+
+	return keys
+		.map((key, index) => {
+			const param = `${key}=${variables[key]}`;
+			return index === keys.length - 1 ? `?${param}` : param;
+		})
+		.join('&');
+};
+
+const useResource = ({
+	fetchDelay = 300,
+	fetchOptions,
+	fetchPolicy = FetchPolicy.NoCache,
+	fetchTimeout = 6000,
+	link,
+	onNetworkStatusChange = () => {},
+	pollInterval = 0,
+	storage = {
+		/**
+		 * This will ensure that we know that the storage
+		 * reference is not external, otherwise the cache
+		 * will persist by the application.
+		 */
+		[SYMBOL_ORIGIN]: true,
+	},
+	fetchRetry = {},
+	storageMaxSize = 20,
+	variables = null,
+}: IResource) => {
+	const [resource, setResource] = useState<any>(null);
+
+	let networkStatus = useRef<NetworkStatus>(NetworkStatus.Unused).current;
+
+	let pollingIntervalId = useRef<null | NodeJS.Timeout>(null).current;
+
+	let retryDelayIntervalId = useRef<null | NodeJS.Timeout>(null).current;
+
+	const firstRenderRef = useRef<boolean>(true);
+
+	const cache = useCache(
+		fetchPolicy,
+		storage,
+		storageMaxSize,
+		link,
+		variables
+	);
+
+	const debouncedVariablesChange = useDebounce(variables, fetchDelay);
+
+	const setNetworkStatus = (status: NetworkStatus) => {
+		networkStatus = status;
+		onNetworkStatusChange(status);
+	};
+
+	const handleRefetch = () => {
+		doFetch();
+		setNetworkStatus(NetworkStatus.Refetch);
+	};
+
+	const cleanRetry = () => {
+		if (retryDelayIntervalId) {
+			clearInterval(retryDelayIntervalId);
+		}
+	};
+
+	const getRetryDelay = (retryAttempts: number) => {
+		const {delay = {}} = fetchRetry;
+		const {initial = 300, jitter = true} = delay;
+
+		const baseDelay = jitter ? initial : initial / 2;
+
+		let value = Math.min(baseDelay * 2 ** retryAttempts);
```

Comment:
Shouldn't `Math.min` take at least 2 arguments?

Reply:
Oh yeah, I would add a minimum time that retry can reach, I just forgot to remove it. Thanks